### PR TITLE
update: issue #2763

### LINKED ISSUE #2763: Multiple definitions of `basic_paradigm`
The function `basic_paradigm` is defined in `nilearn._utils.data_gen.py`:

https://github.com/nilearn/nilearn/blob/00d3873fb6cf23d09f8eba6d0d20e905f26d22dd/nilearn/_utils/data_gen.py#L505-L510

But, it is re-defined in other files (with a duration) for testing:

- `nilearn/glm/tests/test_first_level.py`:
https://github.com/nilearn/nilearn/blob/a36c0442fc8eebe401b310dd9799c0f071080e90/nilearn/glm/tests/test_first_level.py#L353-L360

- `nilearn/glm/tests/test_paradigm.py`:
https://github.com/nilearn/nilearn/blob/a36c0442fc8eebe401b310dd9799c0f071080e90/nilearn/glm/tests/test_paradigm.py#L16-L23

- `nilearn/glm/tests/test_dmtx.py`:
https://github.com/nilearn/nilearn/blob/a36c0442fc8eebe401b310dd9799c0f071080e90/nilearn/glm/tests/test_dmtx.py#L48-L55

It would be better to remove those and import `basic_paradigm` from `_utils.data_gen.py`.

### PR DESCRIPTION
**Pull request by jithenshriyan**
Replaced rewritten functions in multiple files by importing the original function.

Closes #2763.

Changes proposed in this pull request:

- Import `basic_paradigm` function from original file to applicable files 
- Delete rewritten functions in applicable files

### CODE REVIEW

**Comment by jithenshriyan** [comment]
I see an error in my syntax. I'm going to fix it and submit another PR.

Sorry, I'm a new contributor.

File: nilearn/glm/tests/test_dmtx.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
 
 from nibabel.tmpdirs import InTemporaryDirectory
 from numpy.testing import (
-    assert_almost_equal, assert_array_equal, assert_array_almost_equal,
+    assert_almost_equal, assert_array_equal, sassert_array_almost_equal,
```
Comment (NicolasGensollen) [inline nilearn/glm/tests/test_dmtx.py:15]: typo here

File: nilearn/glm/tests/test_dmtx.py [hunk 2]
```diff
@@ -185,7 +177,7 @@ def test_convolve_regressors():
 
 
 def test_design_matrix1():
-    # basic test based on basic_paradigm and glover hrf
+    # basic test based on basic_paradigm() and glover hrf
```
Comment (NicolasGensollen) [inline nilearn/glm/tests/test_dmtx.py:180]: Could you undo these parentheses additions?

File: nilearn/glm/tests/test_first_level.py [hunk 3]
```diff
@@ -395,24 +397,14 @@ def test_fmri_inputs():
...
-
 def test_first_level_design_creation():
     # Test processing of FMRI inputs
     with InTemporaryDirectory():
         shapes = ((7, 8, 9, 10),)
         mask, FUNCFILE, _ = write_fake_fmri_data_and_design(shapes)
         FUNCFILE = FUNCFILE[0]
         func_img = load(FUNCFILE)
-        # basic test based on basic_paradigm and glover hrf
+        # basic test based on basic_paradigm() and glover hrf
```
Comment (NicolasGensollen) [inline nilearn/glm/tests/test_first_level.py:407]: Same comment for this file

File: nilearn/glm/tests/test_paradigm.py [hunk 4]
```diff
@@ -6,21 +6,14 @@
 """
 
 import os
-import pytest
+# import pytest
 import numpy as np
 import pandas as pd
 from numpy.testing import assert_array_equal
-from nilearn.glm.first_level import check_events
+# from nilearn.glm.first_level import check_events
```
Comment (NicolasGensollen) [inline nilearn/glm/tests/test_paradigm.py:13]: If the import is not needed anymore, it is better to just remove it

**Review by NicolasGensollen — CHANGES_REQUESTED** [review]
Thanks for working on this @jithenshriyan ! :+1: 
I enabled the CI to run on your PR, and tests are failing because of a typo (see one of my comments below).
One thing you can do to spot these issues before pushing is to run the tests locally. To run all tests, just run at the root of your local nilearn fork:

```
$ pytest -v --pyargs nilearn
```

To run a specific test file (for example `test_masking.py` here):

```
$ pytest -v nilearn/tests/test_masking.py
```

**Comment by jithenshriyan** [comment]
Thanks! Definitely, a helpful tip as I am a new contributor. Will get to work on this today.

File: nilearn/glm/tests/test_paradigm.py [hunk 5]
```diff
@@ -6,21 +6,13 @@
 """
 
 import os
-import pytest
+# import pytest
```
Comment (NicolasGensollen) [inline nilearn/glm/tests/test_paradigm.py:9]: please remove the line if the import is not needed

File: nilearn/glm/tests/test_dmtx.py [hunk 6]
```diff
@@ -285,7 +277,7 @@ def test_design_matrix8():
 
 
 def test_design_matrix9():
-    # basic test based on basic_paradigm and FIR
+    # basic test based on basic_paradigm() and FIR
```
Comment (NicolasGensollen) [inline nilearn/glm/tests/test_dmtx.py:280]: Could you undo these parentheses additions?

File: nilearn/glm/tests/test_dmtx.py [hunk 7]
```diff
@@ -350,7 +342,7 @@ def test_design_matrix14():
 
 
 def test_design_matrix15():
-    # basic test based on basic_paradigm, plus user supplied regressors
+    # basic test based on basic_paradigm(), plus user supplied regressors
```
Comment (NicolasGensollen) [inline nilearn/glm/tests/test_dmtx.py:345]: same

File: nilearn/glm/tests/test_first_level.py [hunk 8]
```diff
@@ -443,7 +435,7 @@ def test_first_level_glm_computation():
...
```
Comment (NicolasGensollen) [inline nilearn/glm/tests/test_first_level.py:438]: same

File: nilearn/glm/tests/test_paradigm.py [hunk 9]
```diff
@@ -6,21 +6,13 @@
...
 import numpy as np
 import pandas as pd
 from numpy.testing import assert_array_equal
-from nilearn.glm.first_level import check_events
```
Comment (NicolasGensollen) [inline nilearn/glm/tests/test_paradigm.py:13]: please remove the blank line here

File: nilearn/glm/tests/test_first_level.py [hunk 10]
```diff
@@ -25,6 +25,8 @@
 from nilearn.image import get_data
 from nilearn.input_data import NiftiMasker
```
Comment (NicolasGensollen) [inline nilearn/glm/tests/test_first_level.py:27]: please remove this blank line

**Review by NicolasGensollen — CHANGES_REQUESTED** [review]
Thanks for addressing my reviews @jithenshriyan !
I left a few small comments below.
For now you are getting a lot of test failures since the function `basic_paradigm` that you are importing does not return a dataframe with a duration column.

**Comment by jithenshriyan** [comment]
Hi @NicolasGensollen, is there anything else you would like me to change or add before closing the PR? Thanks!

**Comment by NicolasGensollen** [comment]
Hi @jithenshriyan 
There are still a lot of tests failing. I think most of them are failing because the column duration is missing from the dataframe you return.

**Comment by jithenshriyan** [comment]
Interesting. I hadn't written the function myself, was merely just moving it. 

I'd be happy to fix it though!

```python3
def basic_paradigm():
    conditions = ['c0', 'c0', 'c0', 'c1', 'c1', 'c1', 'c2', 'c2', 'c2']
    onsets = [30, 70, 100, 10, 30, 90, 30, 40, 60]
    events = pd.DataFrame({'trial_type': conditions,
                           'onset': onsets})
    return events
```
It seems in the `basic_paradigm()` function we haven't defined a column named duration. It's written to return just conditions and onsets. 

Do you think making these changes would be sufficient?

```python3
def basic_paradigm():
    duration = np.full(20, 0.5)
    conditions = ['c0', 'c0', 'c0', 'c1', 'c1', 'c1', 'c2', 'c2', 'c2']
    onsets = [30, 70, 100, 10, 30, 90, 30, 40, 60]
    events = pd.DataFrame({'trial_type': conditions,
                           'onset': onsets, 'duration': duration})
    return events
```

**Comment by NicolasGensollen** [comment]
>Interesting. I hadn't written the function myself, was merely just moving it.

Yes, these functions were very similar but not exactly the same since the duration column was missing in the one from `data_gen`, probably because this column wasn't used or because the function itself was never used. In any case, having only one of these functions will be better.

>Do you think making these changes would be sufficient?

The code you propose won't work because the duration array will be of length 20 instead of 9, which will fail at dataframe creation. 
I would suggest using the code from one of the deleted functions where `duration = 1 * np.ones(9)`.

**Comment by NicolasGensollen** [comment]
Anything I can do to help @jithenshriyan ? 
I don't think there is much missing here.

**Comment by jithenshriyan** [comment]
Sorry @NicolasGensollen, been a busy week. I'll get to work on it sometime this evening or over the weekend.

Appreciate your help. Thanks!

**Comment by NicolasGensollen** [comment]
@jithenshriyan no worries, there is no rush. 
I was just checking if you were having some issues and if I could be of any help.

**Comment by jithenshriyan** [comment]
Would this be fine then? 

```python
def basic_paradigm(): 
     conditions = ['c0', 'c0', 'c0', 'c1', 'c1', 'c1', 'c2', 'c2', 'c2'] 
     onsets = [30, 70, 100, 10, 30, 90, 30, 40, 60] 
     durations = 1 * np.ones(9) 
     events = pd.DataFrame({'trial_type': conditions, 
                            'onset': onsets, 
                            'duration': durations}) 
     return events 
```

**Comment by jithenshriyan** [comment]
I got it from the original ```basic_paradigm()``` in ```test_first_level.py``` where ```duration = 1 * np.ones(9)```

**Comment by bthirion** [comment]
> Would this be fine then?

LGTM, thx !

File: nilearn/glm/tests/test_paradigm.py [hunk 11]
```diff
@@ -10,18 +10,7 @@
...
```
Comment (NicolasGensollen) [inline nilearn/glm/tests/test_paradigm.py:13]: You need to keep this import

**Review by NicolasGensollen — CHANGES_REQUESTED** [review]
Thanks @jithenshriyan !
You are getting some errors because you removed an import in one of the test file.
Could you also merge/rebase master?

**Comment by ymzayek** [comment]
> Thx @jithenshriyan ! @ymzayek is it OK to rebase it on main ? Best,

Yes I will give it a go

**Pull request closed without merging**